services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -683,11 +688,35 @@ protected boolean isNewRequest(AuthenticationSessionModel authSession, ClientMod
     @NoCache
     @Consumes({"application/soap+xml",MediaType.TEXT_XML})
     public Response soapBinding(InputStream inputStream) {
-        SamlEcpProfileService bindingService = new SamlEcpProfileService(realm, event);
+        SamlEcpProfileService bindingService = new SamlEcpProfileService(realm, event, knownPorts, knownProtocols);
 
         ResteasyProviderFactory.getInstance().injectProperties(bindingService);
 
         return bindingService.authenticate(inputStream);
     }
 
+    private boolean isValidDestination(URI destination) {
+        if (destination == null) {
+            return false;
+        }
+
+        URI expected = uriInfo.getAbsolutePath();
+
+        if (Objects.equals(expected, destination)) {
+            return true;
+        }
+
+        Integer portByScheme = knownPorts.get(expected.getScheme());
+        if (expected.getPort() < 0 && portByScheme != null) {
+            return Objects.equals(uriInfo.getRequestUriBuilder().port(portByScheme).build(), destination);
```

Comment:
It should be 
return Objects.equals(uriInfo.getAbsolutePathBuilder().port(portByScheme).build(), destination); 

Reason : uriInfo.getRequestUriBuilder() return the URIBuilder object initialized with request uri which would contain the SAMLrequest query parameter in the URL but the destination does not have a query parameter hence it will fail.

uriInfo.getAbsolutePathBuilder() will return the builder initialized with absolute path (query param will be dropped here). Hence it will match with the destination URL.